SAMZA-2404: [SEP-22] Container Placement Handler for dispatching container placement messages between metastore and JobCoordinator

Sanil15 opened the pull request:
**Design:** [SEP-22: Container Placements in Samza](https://cwiki.apache.org/confluence/display/SAMZA/SEP-22%3A+Container+Placements+in+Samza)

**Changes:** : Following abilities are added as a part of this PR:

1. Container placement handler which is a stateless handler that periodically dispatches messages between the Metadata store and the Container Placement Service APIs running as a part of Job Coordinator
2. ContainerPlacementRequestMessage & ContainerPlacementResponseMessage are two POJO used to encapsulate request & response messages for container placement actions
3. Custom ObjectMapper for serializing and deserializing these messages is added as a part of ContainerPlacementMessageObjectMapper

**Tests:** 
1. Unit tests are added to test message serialization & deserialization
2. Unit tests to test ContainerPlacementHandler
2. Manual testing has been done to see reads / writes from meta-store

**API Changes:**: 
1. Adds two new public exposed API class: ContainerPlacementRequestMessage & ContainerPlacementResponseMessage in samza-api
2. ContainerPlacementRequestMessage are messages used to issue container placement actions, one can write these messages to meta-store using Meta-store writer 
3. ContainerPlacementResponseMessage are messages used as a response to container placement actions, Job Coordinator writes these messages to Meta-store so external controllers can read response to the current actions

**Upgrade Instructions:** Existing jobs with a persisted metastore will be able to ignore these newly populated messages because they have a different namespace

**Usage Instructions:** An external controller / tool (ex: CoordinatorStreamWriter) would be able to issue such actions by writing to metastore (currently: CoordinatorStream)

Sanil15 commented:
@dnishimura @bharathkk @shanthoosh @rmatharu Please have a look

rmatharu commented on samza-core/src/main/java/org/apache/samza/clustermanager/ContainerProcessManager.java:
```diff
@@ -97,6 +100,8 @@
   // The ContainerManager manages control actions for both active & standby containers
   private final ContainerManager containerManager;
+  private ContainerPlacementHandler containerPlacementHandler;
+  private Thread containerPlacementHandlerThread;
```
Please use ExecutorService executorService = Executors.newSingleThreadExecutor(

rmatharu replied:
Maybe we can use the 
```afterExecute(Runnable r, Throwable t)``` method of ExecutorService?

Sanil15 replied:
nope I have to use the semaphore for propagating the exception from CPH thread to main thread

rmatharu commented on samza-core/src/main/java/org/apache/samza/clustermanager/ContainerProcessManager.java:
```diff
@@ -185,18 +196,22 @@ public ContainerProcessManager(Config config, SamzaApplicationState state, Metri
   public boolean shouldShutdown() {
-    LOG.debug("ContainerProcessManager state: Completed containers: {}, Configured containers: {}, Are there too many failed containers: {}, Is allocator thread alive: {}",
-      state.completedProcessors.get(), state.processorCount, jobFailureCriteriaMet ? "yes" : "no", allocatorThread.isAlive() ? "yes" : "no");
+    LOG.debug(
```
Maybe this should be  LOG.info

rmatharu commented on samza-core/src/main/java/org/apache/samza/clustermanager/container/placement/ContainerPlacementHandler.java:
```diff
@@ -0,0 +1,287 @@
+   * to issue a request to JobCoordinator
+   * @param message
+   */
+  public void writeContainerPlacementResponseMessage(ContainerPlacementResponseMessage message) {
```
Is this intended for use by the client?

rmatharu commented on samza-core/src/main/java/org/apache/samza/clustermanager/container/placement/ContainerPlacementHandler.java:
```diff
@@ -0,0 +1,287 @@
+    Preconditions.checkState(isRunning, "Underlying metadata store not available");
+    Preconditions.checkNotNull(processorId, "processorId cannot be null");
+
+    byte[] messageBytes = requestStore.get(processorId);
```
Do you know what would the actual key for a request/response message look like?

rmatharu commented on samza-core/src/main/java/org/apache/samza/clustermanager/container/placement/ContainerPlacementMessageObjectMapper.java:
```diff
@@ -0,0 +1,113 @@
+        message = new ContainerPlacementResponseMessage(uuid, applicationId, processorId, destinationHost, requestExpiry,
+            statusCode, responseMessage);
+      }
+      return message;
```
Could we add this to SamzaObjectMapper and use SamzaObjectMapper instead?

dnishimura replied:
Yup SamzaObjectMapper is super bloated. IMO, decoupling is good, especially in this specific domain.

rmatharu replied:
The reason its bloated is because the Mixins and custom serializers and deserializers have been defined in the same file. 
You can define yours in separate files but link them in SamzaObjectMapper since this is a part of Samza.

Sanil15 replied:
"You can define yours in separate files but link them in SamzaObjectMapper since this is a part of Samza."

SamzaObjectMapper class is only used for JobModel serde, I think maybe that is the reason @dnishimura also did not link Startpoint object mappers with it.

rmatharu commented on samza-core/src/main/java/org/apache/samza/config/ClusterManagerConfig.java:
```diff
@@ -113,6 +113,12 @@
+   * The container placement handler sleeps for a configurable time before checking for now ContainerPlacement messages
+   */
+  public static final String CLUSTER_MANAGER_CONTAINER_PLACEMENT_HANDLER_SLEEP_MS = "cluster-manager.container-placement-handler.sleep.interval.ms";
+  private static final int DEFAULT_CLUSTER_MANAGER_CONTAINER_PLACEMENT_HANDLER_SLEEP_MS = 5000;
```
Do we expect different jobs to tune this timeout differently?
If not, i'd suggest removing this config

Sanil15 replied:
No for the first pass we can read defaults!

dnishimura commented on samza-api/src/main/java/org/apache/samza/container/placement/ContainerPlacementMessage.java:
```diff
@@ -0,0 +1,144 @@
+    return statusCode;
+  }
+
+  public Long getRequestExpiry() {
```
Please document the time unit (seconds or millis)

dnishimura commented on samza-core/src/main/java/org/apache/samza/clustermanager/ContainerProcessManager.java:
```diff
@@ -242,6 +259,8 @@ public void stop() {
     // Shutdown allocator thread
     containerAllocator.stop();
+    // Shutdown container placement handler thread
+    containerPlacementHandler.stop();
```
Would you need the similar thread join pattern for the container placement handler thread , similar to the container allocator thread?

dnishimura commented on samza-core/src/main/java/org/apache/samza/clustermanager/ContainerProcessManager.java:
```diff
@@ -562,6 +582,17 @@ void onResourceCompletedWithUnknownStatus(SamzaResourceStatus resourceStatus, St
+   * @param requestMessage request containing details of the desited container placement action
+   */
+  public void registerContainerPlacementAction(ContainerPlacementRequestMessage requestMessage) {
+    // Call the ContainerManager#registerContainerPlacementAction
```
Is this a `TODO:` for a follow on PR?

rmatharu commented on samza-core/src/main/java/org/apache/samza/clustermanager/ContainerProcessManager.java:
```diff
@@ -97,6 +100,8 @@
   // The ContainerManager manages control actions for both active & standby containers
   private final ContainerManager containerManager;
+  private ContainerPlacementHandler containerPlacementHandler;
+  private Thread containerPlacementHandlerThread;
```
final

Sanil15 commented:
Addressing @prateekm feedback & adding the following changes
1. Changed how ContainerPlacementHandler wires in JC, wiring CPH with ClusterBasedJobCoordinator
2. Refactored meta-store read/write utils from ContainerPlacementHandler to a separate entity ContainerPlacementUtil, wired it to ClusterBasedJobCoordinator
3. ContainerPlacementUtil is gonna be wired into ContainerPlacementService in a follow-up PR and can be used by outside clients to request container placement actions.
4. Metastore key for Container placement actions is UUID + Message type (req/res) and both request & response messages now belong to the same Namespace in metastore
5. Documentation & other improvements

dnishimura commented on samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java:
```diff
@@ -316,6 +345,19 @@ private void onShutDown() {
+      LOG.info("Stopped container placement handler thread");
+      containerPlacementUtil.stop();
+    } catch (InterruptedException ie) {
+      LOG.error("Container Placement handler thread join threw an interrupted exception", ie);
```
This error log assumes `containerPlacementUtil.stop()` will never throw an exception. Is that true?

rmatharu replied:
+1, perhaps use finally to do a clean guaranteed cleanup?

dnishimura commented on samza-core/src/main/java/org/apache/samza/clustermanager/container/placement/ContainerPlacementMessageObjectMapper.java:
```diff
@@ -0,0 +1,119 @@
+        containerPlacementMessageMap.put("subType", ContainerPlacementResponseMessage.class.getSimpleName());
+        containerPlacementMessageMap.put("responseMessage",
+            ((ContainerPlacementResponseMessage) value).getResponseMessage());
+      }
```
Would you want to add an else check for unexpected subtypes of `ContainerPlacementMessage`? Also may apply to the `deserialize` method below.

dnishimura commented on samza-core/src/test/java/org/apache/samza/clustermanager/container/placement/TestContainerPlacementObjectMapper.java:
```diff
@@ -0,0 +1,75 @@
+import org.codehaus.jackson.map.ObjectMapper;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
```
nit: use non-wildcard imports.

rmatharu commented on samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java:
```diff
@@ -150,6 +152,13 @@
+  /**
+   * Container placement action dispatcher and util
+   */
+  private final ContainerPlacementUtil containerPlacementUtil;
```
Either rename this to ContainerPlacementMetastoreReader or make all its methods static so that it is really a Util. 
A Util should not be instantiated like this and have a complete lifecycle.

kw2542 replied:
+1

rmatharu commented on samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java:
```diff
@@ -150,6 +152,13 @@
+   */
+  private final ContainerPlacementUtil containerPlacementUtil;
+  private final ContainerPlacementHandler containerPlacementHandler;
+  private final Thread containerPlacementHandlerThread;
```
// containerPlacement thread that reads and writes placement-requests and responses to the metastore

rmatharu commented on samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java:
```diff
@@ -264,9 +279,15 @@ public void run() {
 
+      // ContainerPlacementHandler thread has to start after the cpm is started
+      LOG.info("Starting the container placement handler thread");
+      containerPlacementUtil.start();
```
see comment above

rmatharu commented on samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java:
```diff
@@ -264,9 +279,15 @@ public void run() {
+      containerPlacementUtil.start();
+      containerPlacementHandlerThread.start();
+
       boolean isInterrupted = false;
```
should this be volatile?

rmatharu commented on samza-core/src/main/java/org/apache/samza/clustermanager/container/placement/ContainerPlacementHandler.java:
```diff
@@ -0,0 +1,84 @@
+   * {@link org.apache.samza.clustermanager.ContainerManager} and {@link org.apache.samza.clustermanager.ContainerAllocator}
+   */
+  private final ContainerProcessManager containerProcessManager;
+  private final ContainerPlacementUtil containerPlacementUtil;
```
Dont quite understand why both ClusterBasedJC and this class hold on to ContainerPlacementUtil

rmatharu commented on samza-core/src/main/java/org/apache/samza/clustermanager/container/placement/ContainerPlacementHandler.java:
```diff
@@ -0,0 +1,84 @@
+      } catch (Exception e) {
+        LOG.error(
+            "Got unknown Exception while registering ContainerPlacement actions in ContainerPlacementHandler thread.",
+            e);
```
indent

rmatharu commented on samza-core/src/main/java/org/apache/samza/clustermanager/container/placement/ContainerPlacementUtil.java:
```diff
@@ -0,0 +1,273 @@
+  /**
+   * Deletes all {@link ContainerPlacementMessage}
+   */
+  public void deleteAllContainerPlacementMessages() {
```
This method is not invoked from anywhere, then why do we need it?

rmatharu commented on samza-core/src/main/java/org/apache/samza/clustermanager/container/placement/ContainerPlacementUtil.java:
```diff
@@ -0,0 +1,273 @@
+    Preconditions.checkState(!stopped, "Underlying metadata store not available");
+    Preconditions.checkNotNull(uuid, "uuid cannot be null");
+    containerPlacementMessageStore.delete(toContainerPlacementMessageKey(uuid, ContainerPlacementResponseMessage.class));
+  }
```
If these two are invoked from within the class, should they be private?

rmatharu commented on samza-core/src/test/java/org/apache/samza/clustermanager/TestContainerAllocatorWithoutHostAffinity.java:
```diff
@@ -54,7 +54,8 @@
 
-  private final SamzaApplicationState state = new SamzaApplicationState(jobModelManager);
+  private final SamzaApplicationState state =
+      new SamzaApplicationState(jobModelManager);
```
undo

kw2542 commented on samza-api/src/main/java/org/apache/samza/container/placement/ContainerPlacementMessage.java:
```diff
@@ -0,0 +1,152 @@
+    Preconditions.checkNotNull(destinationHost, "destinationHost cannot be null");
+    Preconditions.checkNotNull(statusCode, "statusCode cannot be null");
+    Preconditions.checkNotNull(timestamp, "Timestamp of the message cannot be null");
+    this.uuid = uuid;
```
We can do null check and assignment in one line:

this.uuid = Preconditions.checkNotNull(uuid, "uuid cannot be null");

same for others.

kw2542 replied:
We can use Preconditions.checkNotNull but in one line, e.g.

this.uuid = Preconditions.checkNotNull(uuid, "uuid cannot be null");

Sanil15 replied:
I misread this, sure changing it

kw2542 commented on samza-api/src/main/java/org/apache/samza/container/placement/ContainerPlacementMessage.java:
```diff
@@ -0,0 +1,152 @@
+    ContainerPlacementMessage message = (ContainerPlacementMessage) o;
+    return getUuid().equals(message.getUuid()) && getDeploymentId().equals(message.getDeploymentId())
+        && getProcessorId().equals(message.getProcessorId()) && getDestinationHost().equals(
+        message.getDestinationHost());
```
is it expected that we are ignoring requestExpiry, statusCode and timestamp when comparing? Same for other classes.

kw2542 replied:
I understand that in reality, each message will have its own UUID, but in terms of comparing java object itself, should we compare all fields? It may be very confusing if two ContainerPlacementMessage equals but their status code is different.

Sanil15 replied:
I see your point will change it

rmatharu replied:
+1, Please make sure all POJO fields are in the equals and hashCode if they can be.

kw2542 commented on samza-api/src/main/java/org/apache/samza/container/placement/ContainerPlacementRequestMessage.java:
```diff
@@ -0,0 +1,62 @@
+    sb.append(", requestExpiry=").append(requestExpiry);
+    sb.append(", statusCode=").append(statusCode);
+    sb.append(", timestamp=").append(timestamp);
+    sb.append('}');
```
How about we have a static class variable for this format, and toString() just do

String.format(FORMAT, uuid, processorId, deploymentId .....)

kw2542 replied:
I will leave this to your judgement, both are fine to me.

kw2542 commented on samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java:
```diff
@@ -264,9 +279,15 @@ public void run() {
 
-      while (!containerProcessManager.shouldShutdown() && !checkAndThrowException() && !isInterrupted) {
+      while (!containerProcessManager.shouldShutdown() && !checkAndThrowException() && !isInterrupted
+          && checkContainerPlacementHandlerThreadIsAlive()) {
```
nit: I personally prefer to have a new line for each condition if they can not fit in one line:

!containerProcessManager.shouldShutdown()
&& !checkAndThrowException()
&& !isInterrupted
&& checkContainerPlacementHandlerThreadIsAlive()

kw2542 commented on samza-core/src/main/java/org/apache/samza/clustermanager/container/placement/ContainerPlacementMessageObjectMapper.java:
```diff
@@ -0,0 +1,119 @@
+        if (code.name().equals(node.get("statusCode").getTextValue())) {
+          statusCode = code;
+        }
+      }
```
It is not easy to read/understand, can we do:

statusCode = ContainerPlacementMessage.StatusCode.valueOf(node.get("statusCode").getTextValue())

kw2542 commented on samza-core/src/main/java/org/apache/samza/clustermanager/container/placement/ContainerPlacementUtil.java:
```diff
@@ -0,0 +1,273 @@
+    Preconditions.checkNotNull(deploymentId, "deploymentId should not be null");
+    Preconditions.checkNotNull(processorId, "processorId not be null");
+    Preconditions.checkNotNull(destinationHost, "destinationHost should not be null");
+    Preconditions.checkNotNull(timestamp, "timestamp should not be null");
```
Constructor of ContainerPlacementRequestMessage handles null checks, do we need to do another null check here?

kw2542 commented on samza-core/src/main/java/org/apache/samza/clustermanager/container/placement/ContainerPlacementMessageObjectMapper.java:
```diff
@@ -0,0 +1,119 @@
+    objectMapper.registerSubtypes(new NamedType(ContainerPlacementResponseMessage.class));
+    objectMapper.registerSubtypes(new NamedType(ContainerPlacementRequestMessage.class));
+    return objectMapper;
+  }
```
Can we have a static ObjectMapper? It is thread safe and we are not changing its configuration either.

Sanil15 replied:
I suppose you mean to have a static object Mapper declared here: https://github.com/apache/samza/pull/1227/files#diff-20fafbc969e1be36e38d63017cbfc77aR58?

kw2542 replied:
Yes and no, we can change that to be a static field or getObjectMapper() can return a static singleton as well.

Since getObjectMapper() takes no argument, so it is returning a object mapper with the same configuration, we could make getObjectMapper() to return the same instance of ObjectMapper instead of relying the caller to have a static variable.

Sanil15 replied:
I see sure since we are not changing the configuration it can be a singleton

rmatharu commented on samza-core/src/test/java/org/apache/samza/clustermanager/container/placement/TestContainerPlacementObjectMapper.java:
```diff
@@ -0,0 +1,76 @@
+    assertTrue(message instanceof ContainerPlacementRequestMessage);
+    ContainerPlacementRequestMessage deserializedRequest = (ContainerPlacementRequestMessage) message;
+    assertEquals(requestMessage, deserializedRequest);
+    assertEquals(deserializedRequest.getStatusCode(), ContainerPlacementMessage.StatusCode.CREATED);
```
Since ContainerPlacementMessage.equals  includes statusCode, this step is not needed.

Sanil15 replied:
sure

rmatharu commented on samza-core/src/main/java/org/apache/samza/config/ClusterManagerConfig.java:
```diff
@@ -227,7 +227,6 @@ public int getContainerRetryWindowMs() {
     }
   }
 
-
```
Remove change?

Sanil15 replied:
sure

rmatharu commented on samza-core/src/main/java/org/apache/samza/clustermanager/container/placement/ContainerPlacementRequestAllocator.java:
```diff
@@ -0,0 +1,84 @@
+        Thread.currentThread().interrupt();
+      } catch (Exception e) {
+        LOG.error(
+            "Got unknown Exception while reading ContainerPlacementRequestMessage in ContainerPlacementRequestAllocator thread",
```
Well, if we know the Exception so its not really "unknown" :-)

Sanil15 replied:
removed "unknown"
